src/plugins/WebOfTrust/util/TickerDelayedBackgroundJob.java

```diff
@@ -0,0 +1,363 @@
+            assert(thread == null) : "having job thread while going to TERMINATED state";
+            assert(waitingTickerJob != null) : "in WAITING state but no ticker job";
+            // Remove the scheduled job from the ticker on a best-effort basis.
+            ticker.removeQueuedJob(waitingTickerJob);
+            waitingTickerJob = null;
+        } else {
+            assert(state == JobState.IDLE) : "going to TERMINATED state from illegal state";
+            assert(thread == null) : "having job thread while going to TERMINATED state";
+            assert(waitingTickerJob == null) : "going to TERMINATED with waiting ticker job";
+        }
+        state = JobState.TERMINATED;
+        // Notify all threads waiting in waitForTermination()
+        notifyAll();
+    }
+
+    /**
+     * Indicates whether this job may start and sets the state to {@code RUNNING}
+     * accordingly. If this job is {@code TERMINATED}, it may not start, if it is {@code
+     * WAITING}, it may start.
+     * @return {@code true} if the job may start
+     */
+    private synchronized boolean jobStarted() {
+        if (state == JobState.TERMINATED) {
+            return false;
+        }
+        toRUNNING();
+        return true;
+    }
+
+    /**
+     * Finishes the job by either enqueueing itself again in {@code WAITING} (if there has been a
+     * trigger since the start of the last job), waiting for a trigger in {@code IDLE} or going
+     * to the {@code TERMINATED} state if we were previously {@code TERMINATING}.
+     */
+    private synchronized void jobFinished() {
+        if (state == JobState.TERMINATED) {
+            return;
+        }
+        if (state == JobState.TERMINATING) {
+            toTERMINATED();
+            return;
+        }
+        toIDLE();
+        if (nextExecutionTime != NO_EXECUTION) {
+            long delay = nextExecutionTime - System.currentTimeMillis();
+            enqueueWaitingTickerJob(delay);
+        }
+    }
+
+    /**
+     * A wrapper for jobs. After the job finishes, either goes to an {@code IDLE} state or enqueues
+     * its own next run {@code WAITING} (implementation in {@link #jobFinished()}).
+     */
+    private class DelayedBackgroundRunnable implements PrioRunnable {
+        private final Runnable job;
+
+        DelayedBackgroundRunnable(Runnable job) {
+            this.job = job;
+        }
+
+        @Override
+        public void run() {
+            try {
+                if (jobStarted()) {
+                    job.run();
+                }
+            } finally {
+                jobFinished();
+            }
+        }
+
+        @Override
+        public int getPriority() {
+            if (job instanceof PrioRunnable) {
+                return ((PrioRunnable)job).getPriority();
+            }
+            return NativeThread.NORM_PRIORITY;
+        }
+    }
+
+    /**
+     * For testing purposes. Returns the internal job state.
+     */
+    synchronized JobState getState() {
```

Comment:
What is possible is that the contents get cached under certain circuumstances. To be absolutely safe, the removal of the "synchronized" could be replaced with making the variable volatile to disable caching.